### PR TITLE
[libc][math][c23] Add fsub{,l,f128} and remainderf128 C23 math functions

### PR DESCRIPTION
**Pull request by aaryanshukla**

### CODE REVIEW

**Review by michaelrj-google — COMMENTED** [review]
is seems like there should be a test for remainderf128 as well

**Comment by aaryanshukla** [comment]
@michaelrj-google None of the remainder functions seem to have tests yet. @overmighty

**Comment by aaryanshukla** [comment]
@overmighty was able to get the MPFR tests, will look into ddivl and the rounding issues there. Let me know if this looks good.

**Comment by aaryanshukla** [comment]
@aaryanshukla @michaelrj-google should I leave remainder tests out then if we do not have any?

**Comment by overmighty** [comment]
You should probably avoid putting to-do notes in the PR description so that they don't end up in the merge/squash commit message, as in https://github.com/llvm/llvm-project/commit/8f33f1d829bfa8ab7af5fc60b413cba8994c33a1.

**Comment by overmighty** [comment]
I forgot that `remainder*` functions are pretty much the same as `remquo*`, and we do have tests for those, so I guess that's why we don't have tests for `remainder*`.

File: libc/utils/MPFRWrapper/MPFRUtils.cpp [hunk 1]
```diff
@@ -1086,9 +1086,14 @@ template void
...
                                            const TernaryInput<long double> &,
                                            long double, double, RoundingMode);
+template void explain_ternary_operation_one_output_error(
+    Operation, const TernaryInput<double> &, float, double, RoundingMode);
+template void explain_ternary_operation_one_output_error(
+    Operation, const TernaryInput<long double> &, float, double, RoundingMode);
 
 template void explain_ternary_operation_one_output_error(
     Operation, const TernaryInput<long double> &, double, double, RoundingMode);
+
```
Comment (overmighty) [inline libc/utils/MPFRWrapper/MPFRUtils.cpp:1096]: Nit: sorting.

You can apply this with `git apply path/to/patch.diff`.

```diff
diff --git a/libc/utils/MPFRWrapper/MPFRUtils.cpp b/libc/utils/MPFRWrapper/MPFRUtils.cpp
index 1695354565f5..ec81ea2a9d7d 100644
--- a/libc/utils/MPFRWrapper/MPFRUtils.cpp
+++ b/libc/utils/MPFRWrapper/MPFRUtils.cpp
@@ -1080,20 +1080,18 @@ void explain_ternary_operation_one_output_error(
 
 template void explain_ternary_operation_one_output_error(
     Operation, const TernaryInput<float> &, float, double, RoundingMode);
-template void explain_ternary_operation_one_output_error(
-    Operation, const TernaryInput<double> &, double, double, RoundingMode);
-template void
-explain_ternary_operation_one_output_error(Operation,
-                                           const TernaryInput<long double> &,
-                                           long double, double, RoundingMode);
 template void explain_ternary_operation_one_output_error(
     Operation, const TernaryInput<double> &, float, double, RoundingMode);
+template void explain_ternary_operation_one_output_error(
+    Operation, const TernaryInput<double> &, double, double, RoundingMode);
 template void explain_ternary_operation_one_output_error(
     Operation, const TernaryInput<long double> &, float, double, RoundingMode);
-
 template void explain_ternary_operation_one_output_error(
     Operation, const TernaryInput<long double> &, double, double, RoundingMode);
-
+template void
+explain_ternary_operation_one_output_error(Operation,
+                                           const TernaryInput<long double> &,
+                                           long double, double, RoundingMode);
 #ifdef LIBC_TYPES_HAS_FLOAT16
 template void explain_ternary_operation_one_output_error(
     Operation, const TernaryInput<float> &, float16, double, RoundingMode);
@@ -1275,15 +1273,14 @@ template bool compare_ternary_operation_one_output(Operation,
                                                    const TernaryInput<double> &,
                                                    double, double,
                                                    RoundingMode);
+template bool compare_ternary_operation_one_output(
+    Operation, const TernaryInput<long double> &, float, double, RoundingMode);
+template bool compare_ternary_operation_one_output(
+    Operation, const TernaryInput<long double> &, double, double, RoundingMode);
 template bool
 compare_ternary_operation_one_output(Operation,
                                      const TernaryInput<long double> &,
                                      long double, double, RoundingMode);
-
-template bool compare_ternary_operation_one_output(
-    Operation, const TernaryInput<long double> &, double, double, RoundingMode);
-template bool compare_ternary_operation_one_output(
-    Operation, const TernaryInput<long double> &, float, double, RoundingMode);
 #ifdef LIBC_TYPES_HAS_FLOAT16
 template bool compare_ternary_operation_one_output(Operation,
                                                    const TernaryInput<float> &,
```

**Comment by overmighty** [comment]
`-DLIBC_USE_NEW_HEADER_GEN=OFF` is still an option, so you should still update libc/spec/stdc.td.

File: libc/utils/MPFRWrapper/MPFRUtils.cpp (hunk 1 above)
Comment (aaryanshukla) [reply libc/utils/MPFRWrapper/MPFRUtils.cpp:1096]: Will first merge fdiv and then match the MPFRwrapper.

File: libc/config/linux/aarch64/entrypoints.txt [hunk 2]
```diff
@@ -693,6 +695,7 @@ if(LIBC_TYPES_HAS_FLOAT128)
     libc.src.math.frexpf128
     libc.src.math.fromfpf128
     libc.src.math.fromfpxf128
+    libc.src.math.fsubf128
     libc.src.math.fsqrtf128
```
Comment (overmighty) [inline libc/config/linux/aarch64/entrypoints.txt:699]: Nit: sorting.

```suggestion
    libc.src.math.fsqrtf128
    libc.src.math.fsubf128
```

File: libc/config/linux/riscv/entrypoints.txt [hunk 3]
```diff
@@ -593,10 +595,12 @@ if(LIBC_TYPES_HAS_FLOAT128)
     libc.src.math.fminimumf128
     libc.src.math.fmodf128
     libc.src.math.fmulf128
+    libc.src.math.remainderf128
```
Comment (overmighty) [inline libc/config/linux/riscv/entrypoints.txt:598]: Nit: sorting.

File: libc/src/math/generic/CMakeLists.txt [hunk 4]
```diff
@@ -2884,6 +2884,43 @@ add_entrypoint_object(
...
 )
 
+add_entrypoint_object(
+  fsub
+  SRCS
+    fsub.cpp
+  HDRS
+    ../fsub.h
+  DEPENDS
+   libc.src.__support.FPUtil.generic.add_sub
```
Comment (overmighty) [inline libc/src/math/generic/CMakeLists.txt:2894]: Nit: whitespace.

```suggestion
  DEPENDS
    libc.src.__support.FPUtil.generic.add_sub
```

File: libc/src/math/generic/CMakeLists.txt [hunk 5]
```diff
@@ -2884,6 +2884,43 @@ add_entrypoint_object(
...
+)
+
+add_entrypoint_object(
+  fsubl
+  SRCS
+    fsubl.cpp
+  HDRS
+    ../fsubl.h
+  DEPENDS
+   libc.src.__support.FPUtil.generic.add_sub
```
Comment (overmighty) [inline libc/src/math/generic/CMakeLists.txt:2906]: Nit: whitespace.

File: libc/src/math/generic/CMakeLists.txt [hunk 6]
```diff
@@ -2884,6 +2884,43 @@ add_entrypoint_object(
...
+
+add_entrypoint_object(
+  fsubf128
+  SRCS
+    fsubf128.cpp
+  HDRS
+    ../fsubf128.h
+  DEPENDS
+   libc.src.__support.macros.properties.types
+   libc.src.__support.FPUtil.generic.add_sub
```
Comment (overmighty) [inline libc/src/math/generic/CMakeLists.txt:2919]: Nit: whitespace.

File: libc/newhdrgen/yaml/math.yaml [hunk 7]
```diff
@@ -794,6 +794,14 @@ functions:
     arguments:
       - type: long double
       - type: long double
+  - name: remainderf128
+    standards:
+      - llvm_libc_ext
```
Comment (overmighty) [inline libc/newhdrgen/yaml/math.yaml:799]: `remainderf128` is a standard C function.

File: libc/spec/llvm_libc_ext.td [hunk 8]
```diff
@@ -100,6 +100,9 @@ def LLVMLibcExt : StandardSpec<"llvm_libc_ext"> {
 
           FunctionSpec<"powi", RetValSpec<DoubleType>, [ArgSpec<DoubleType>, ArgSpec<IntType>]>,
           FunctionSpec<"powif", RetValSpec<FloatType>, [ArgSpec<FloatType>, ArgSpec<IntType>]>,
+
+          GuardedFunctionSpec<"remainderf128", RetValSpec<Float128Type, [ArgSpec<Float128Type>, ArgSpec<Float128Type>], "LIBC_TYPES_HAS_FLOAT128">,
```
Comment (overmighty) [inline libc/spec/llvm_libc_ext.td:104]: `remainderf128` is a standard C function.

`fsubf128` is missing.

**Review by overmighty — CHANGES_REQUESTED** [review]

File: libc/newhdrgen/tests/input/test_small.yaml
Comment (overmighty) [inline libc/newhdrgen/tests/input/test_small.yaml:1]: You should make a separate PR for unrelated newhdrgen YAML cleanups.

File: libc/spec/stdc.td [hunk 9]
```diff
@@ -752,6 +752,9 @@ def StdC : StandardSpec<"stdc"> {
           FunctionSpec<"getpayloadf", RetValSpec<FloatType>, [ArgSpec<FloatPtr>]>,
           GuardedFunctionSpec<"getpayloadf16", RetValSpec<Float16Type>, [ArgSpec<Float16Ptr>], "LIBC_TYPES_HAS_FLOAT16">,
           GuardedFunctionSpec<"getpayloadf128", RetValSpec<Float128Type>, [ArgSpec<Float128Ptr>], "LIBC_TYPES_HAS_FLOAT128">,
+
+          GuardedFunctionSpec<"remainderf128", RetValSpec<Float128Type, [ArgSpec<Float128Type>, ArgSpec<Float128Type>], "LIBC_TYPES_HAS_FLOAT128">,
```
Comment (overmighty) [inline libc/spec/stdc.td:758]: Nit: should be grouped with the other `remainder*` functions.

**Review by overmighty — APPROVED** [review]
LGTM with whitespace nits.
